Docs - k8s - Webhook Certs on Vault (#13441)
hashicorp/consul: website/content/docs/k8s/installation/vault/data-integration/webhook-certs.mdx

---
layout: docs
page_title: Vault as the Webhook Certificate Provider for Consul Controller and Connect Inject on Kubernetes
description: >-
  Configuring the Consul Helm chart to use TLS certificates issued by Vault for the Consul Controller and Connect Inject webhooks.
---

> # Vault as the Controller and Connect Inject Webhook Certificate Provider on Kubernetes
> ## Prerequisites

Reviewer: ```suggestion
# Vault as the Controller and Connect Inject Webhook Certificate Provider on Kubernetes

This topic describes how to configure the Consul Helm chart to use TLS certificates issued by Vault in the Consul controller and connect inject webhooks.

## Prerequisites
```
Per the style guide, we don't want to introduce sections so that headings aren't stacked on top of each other.

> Prior to setting up this data integration between Vault and Consul on Kubernetes, you will need to:  
> 1. Read and complete the steps in the [Systems Integration](/docs/k8s/installation/vault/systems-integration) section of [Vault as a Secrets Backend](/docs/k8s/installation/vault).
> 1. Read the [Data Integration Overview](/docs/k8s/installation/vault/data-integration) section of [Vault as a Secrets Backend](/docs/k8s/installation/vault).
> 1. Configure [Vault as the Server TLS Certificate Provider on Kubernetes](/docs/k8s/installation/vault/data-integration/server-tls) 
> 1. Configure [Vault as the Service Mesh Certificate Provider on Kubernetes](/docs/k8s/installation/vault/data-integration/connect-ca)
> 1. Complete the [Bootstrapping the PKI Engine for Controller and Connect Inject Webhooks](#bootstrapping-the-pki-engine-for-controller-and-connect-inject-webhooks) section.

Reviewer: ```suggestion
Complete the following prerequisites prior to implementing the integration described in this topic:  
* Verify that you have completed the steps described in [Systems Integration](/docs/k8s/installation/vault/systems-integration).
* You should be familiar with the [Data Integration Overview](/docs/k8s/installation/vault/data-integration).
* Configure [Vault as the Server TLS Certificate Provider on Kubernetes](/docs/k8s/installation/vault/data-integration/server-tls) 
* Configure [Vault as the Service Mesh Certificate Provider on Kubernetes](/docs/k8s/installation/vault/data-integration/connect-ca)
* Complete the [Bootstrapping the PKI Engine for Controller and Connect Inject Webhooks](#bootstrapping-the-pki-engine-for-controller-and-connect-inject-webhooks) section.
```
Reformatting per the style guide. Revert back to a numbered list if these steps must be completed in order.

Author: 👍 Thank you.  these have to be done in order.  I took the rest of the suggestions.  🙏

> ### Bootstrapping the PKI Engine for Controller and Connect Inject Webhooks

Reviewer 2: I feel like this section should be moved down into the main section of the article. With it as a prereq, the reader is already pretty far into the article before they get to the "Overview".

Author: Good call.  Adam said the same thing.  I've moved this down after Overview on this page and the other pages in this section.

> First, we need to bootstrap the Vault cluster by enabling and configuring the PKI Secrets Engine to be able to serve TLS certificates for the controller and connect inject webhooks.

Reviewer: ```suggestion
The first step is to bootstrap the Vault cluster. Issue the following commands to enable and configure the PKI Secrets Engine to serve TLS certificates for the controller and connect inject webhooks:
```
"We" should only refer to HashiCorp. Address the reader as "you" or reword as an imperative so that "you" is implied.

* Mount the PKI Secrets Engine for each:

  ```shell-session
  $ vault secrets enable -path=controller pki
  ```

  ```shell-session
  $ vault secrets enable -path=connect-inject pki
  ```

* Tune the engine mounts to enable longer TTL:

  ```shell-session
  $ vault secrets tune -max-lease-ttl=87600h controller
  ```

  ```shell-session
  $ vault secrets tune -max-lease-ttl=87600h connect-inject
  ```

* Generate the root CA for each:

  ```shell-session
  $ vault write -field=certificate controller/root/generate/internal \
          common_name="<RELEASE NAME>-controller-webhook" \
          ttl=87600h
  ```

  ```shell-session
  $ vault write -field=certificate connect-inject/root/generate/internal \
          common_name="<RELEASE NAME>-connect-injector" \
          ttl=87600h

>   ```

Reviewer 2: This is the first time a reader is seeing Release Name in this tutorial. May be worthwhile to give them a quick refresher on the topic.

Author: good call.  thank you. I am changing it to <helm release name> which is used on other pages

> ## Overview

Reviewer: I'm confused as to why the overview is in the middle of the topic instead of at the top.

> In a Consul Helm chart configuration that does not use Vault, webhook-cert-manager normally fulfills the role of ensuring that a valid certificate is updated to the mutatingwebhook configuration of either controller or connect inject to ensure that Kubernetes can communicate with each of these services. 

Reviewer 2: ```suggestion
In a Consul Helm chart configuration that does not use Vault, `webhook-cert-manager` normally fulfills the role of ensuring that a valid certificate is updated to the `mutatingwebhook` configuration of either controller or connect inject to ensure that Kubernetes can communicate with each of these services. 
```

Author: added a few different code styles and also fixed things link `mutatingwebhookconfiguration` shoudl be one word.

When Vault is configured as the controller and connect inject Webhook Certificate Provider on Kubernetes: 
  - webhook-cert-manager is no longer deployed to the cluster.
  - controller and connect inject each get their webhook certificates from its own Vault PKI mount via the injected Vault Agent.
  - controller and connect inject each need to be configured with its own Vault Role that has necessary permissions to receive certificates from its respective PKI mount.

>   - controller and connect inject each locally update their mutatingwebhookconfigurations so that Kubernetes can relay events.

Reviewer 2: Should some of these be formatted as `code`?

>   - Vault will manage certificate rotation and each of these webhooks will seamlessly received rotated certificates.

Reviewer: ```suggestion
  - Vault manages certificate rotation and rotates certificates to each webhook.
```
Using present tense and simplifying the statement. Is this still accurate?

To use Vault as the controller and connect inject Webhook Certificate Provider, we will need to modify the steps outlined in the [Data Integration](/docs/k8s/installation/vault/data-integration) section:

### One time setup in Vault
  1. Create a Vault policy that authorizes the desired level of access to the secret.

### Setup per Consul datacenter

>   1. (Added) Create Vault PKI roles for controller and connect inject each that establish the domains that each is allowed to issue certificates for.

Reviewer 2: I don't understand why "(added)" is here.

Reviewer 2: Oh, I see. Reading back it's showing what is modified from the steps in the Data Integration section. That was not initially clear to me.

Reviewer: It would be clearer to just repeat the steps described in Data Integration with modifications as if we are describing a new process. It's too much to ask users, especially new/novice users, to go back and forth in this way.

  1. Create Vault Kubernetes auth roles that link the policy to each Consul on Kubernetes service account that requires access.
  1. Configure the Vault Kubernetes auth roles in the Consul on Kubernetes helm chart.

## One time setup in Vault

> 

Reviewer: ```suggestion
You only need to perform the following steps once in Vault to ? 
```
I left a question mark because it's not clear what the one time setup is referring to. In any case, we need to introduce the following sections somehow.

Author: getting this data integrated means we integrate the systems.  Then we integrate the data.  When we integrate the data there are things that either need to be done once in vault or have to be done for every single data center you have.

So all of these vault pages try to paint the mental model since we are doing this in reference material and not a tutorial, to describe what you are doing and whether it is only once or is repeated for however many data centers you have.

This Vault stuff is very hard to write up in a reference style and this is the best that we have come up with so far to separate:
- system integration tasks vs data integration tasks
- one time set up vs repeat per data center

We are open to other ways to do this, but these vault docs are more involved in terms of a ton of manual steps than anything else probably in our kubernetes section.  So it has been difficult to get something that makes sense for everyone at first glance.

Author: I've moved this to the below section where it needs to be for each data center.

### Create a Vault policy that authorizes the desired level of access to the secret

> To use Vault to issue controller or connect inject webhook certificates, you will need to create the Vault Policies that will allow either controller or connect inject to access its respective certificate issuing url.

Reviewer: ```suggestion
To use Vault to issue controller or connect inject webhook certificates, you will need to create the Vault policies that will allow either controller or connect inject to access its respective certificate-issuing URL.
```

#### Create Vault Policies for the Controller and Connect Inject Webhook Certificates

-> **Note:** The PKI secret paths referenced by the Vault Policies below will be your `global.secretsBackend.vault.controller.tlsCert.secretName` and `global.secretsBackend.vault.connectInject.tlsCert.secretName` Helm values respectively.

> Next, we will create a policy that allows `["create", "update"]` access to the
> [certificate issuing URL](https://www.vaultproject.io/api/secret/pki#generate-certificate) so Consul controller and connect inject can
> fetch a new certificate/key pair and provide it to the Kubernetes mutatingwebhookconfiguration.
> 

Reviewer: ```suggestion
The next step is to create a policy that allows `["create", "update"]` access to the
[certificate issuing URL](https://www.vaultproject.io/api/secret/pki#generate-certificate) so Consul controller and connect inject can fetch a new certificate/key pair and provide it to the Kubernetes `mutatingwebhookconfiguration`. Issue the following commands to create the policy:

```

```shell-session
$ vault policy write controller-tls-policy - <<EOF
path controller/issue/controller-role {
  capabilities = ["create", "update"]
}
EOF
```

```shell-session
$ vault policy write connect-inject-policy - <<EOF
path connect-inject/issue/connect-inject-role {
  capabilities = ["create", "update"]
}
EOF
```
#### Create Vault Policies for the CA URL

-> **Note:** The PKI secret paths referenced by the Vault Policies below will be your `global.secretsBackend.vault.controller.caCert.secretName` and `global.secretsBackend.vault.connectInject.caCert.secretName` Helm values respectively.

> Next, we will create a policy that allows `["read"]` access to the [CA URL](https://www.vaultproject.io/api/secret/pki#read-certificate),
> this is required for the Consul components to communicate with the Consul servers in order to fetch their auto-encryption certificates.

Reviewer: ```suggestion
Next, create a policy that allows `["read"]` access to the [CA URL](https://www.vaultproject.io/api/secret/pki#read-certificate). The policy is required so that Consul components can communicate with the Consul servers in order to fetch their auto-encryption certificates. Issue the following commands to create the policy:
```

```shell-session
$ vault policy write controller-ca-policy - <<EOF
path controller/cert/ca {
  capabilities = ["read"]
}
EOF
```

```shell-session
$ vault policy write connect-inject-ca-policy - <<EOF
path connect-inject/cert/ca {
  capabilities = ["read"]
}
EOF
```

> ## Setup per Consul datacenter
> ### Create a Vault PKI role that establishes the domains that it is allowed to issue certificates for

Reviewer: ```suggestion
## Setup per Consul datacenter

Perform the following steps for each Consul datacenter.

### Enable domains to issue certificates

Create a Vault PKI role that establishes the domains that are allowed to issue certificates.
```
The H3 is a really long heading. Does this capture the purpose of this subtopic?

Author: Thank you.  I went with `Configure allowed domains for PKI certificates`

> Create a Vault role for the controller PKI engine will set the default certificate issuance parameters:

Reviewer: ```suggestion
Issue the following command to create a Vault role for the controller PKI engine and set the default parameters for issuing certificates:
```

```shell-session
$ vault write controller/roles/controller-role \
    allowed_domains="<Allowed-domains-string>" \
    allow_subdomains=true \
    allow_bare_domains=true \
    allow_localhost=true \
    generate_lease=true \
    max_ttl="720h"
```

> Create a Vault role for the connect inject PKI engine will set the default certificate issuance parameters:

Reviewer: ```suggestion
Issue the following command to create a Vault role for the connect inject PKI engine and set the default parameters for issuing certificates:
```

```shell-session
$ vault write connect-inject/roles/connect-inject-role \
    allowed_domains="<Allowed-domains-string>" \
    allow_subdomains=true \
    allow_bare_domains=true \
    allow_localhost=true \
    generate_lease=true \
    max_ttl="720h"
```

To generate the `<Allowed-domains-string>` for each use the following script as a template:

```shell-session
#!/bin/sh

# NAME is set to either the value from `global.name` from your Consul K8s value file, or your $HELM_RELEASE_NAME-consul
export NAME=consulk8s
# NAMESPACE is where the Consul on Kubernetes is installed
export NAMESPACE=consul
# DATACENTER is the value of `global.datacenter` from your Helm values config file
export DATACENTER=dc1

echo allowed_domains_controller=\"${NAME}-controller-webhook,${NAME}-controller-webhook.${NAMESPACE},${NAME}-controller-webhook.${NAMESPACE}.svc,${NAME}-controller-webhook.${NAMESPACE}.svc.cluster.local\""

echo allowed_domains_connect_inject=\"${NAME}-connect-injector,${NAME}-connect-injector.${NAMESPACE},${NAME}-connect-injector.${NAMESPACE}.svc,${NAME}-connect-injector.${NAMESPACE}.svc.cluster.local\""
```

> ### Create a Vault auth roles that link the policy to each Consul on Kubernetes service account that requires access

Reviewer: ```suggestion
### Link the Vault policy to Consul workloads
Create a Vault auth roles that link the policy to each Consul workload on Kubernetes service account that requires access.
```
Is it workloads? It wasn't clear what "each Consul" meant. Cluster?

Author: yes.  thank you.  updated in several places.

Finally, Kubernetes auth roles need to be created for controller and connect inject webhooks.

Role for Consul controller webhooks:
```shell-session
$ vault write auth/kubernetes/role/controller-role \
    bound_service_account_names=<Consul controller service account> \
    bound_service_account_namespaces=<Consul installation namespace> \
    policies=controller-ca-policy \
    ttl=1h
```

To find out the service account name of the Consul controller,
you can run:

```shell-session
 $ helm template --release-name ${RELEASE_NAME} --show-only templates/controller-serviceaccount.yaml hashicorp/consul
```

Role for Consul connect inject webhooks:

```shell-session
$ vault write auth/kubernetes/role/connect-inject-role \
    bound_service_account_names=<Consul connect inject service account> \
    bound_service_account_namespaces=<Consul installation namespace> \
    policies=connect-inject-ca-policy \
    ttl=1h
```

To find out the service account name of the Consul connect inject, use the command below.
```shell-session
 $ helm template --release-name ${RELEASE_NAME} --show-only templates/connect-inject-serviceaccount.yaml hashicorp/consul
```

> ### Configure the Vault Kubernetes auth role in the Consul on Kubernetes helm chart
> 
> Now that we've configured Vault, you can configure the Consul Helm chart to
> use the Server TLS certificates from Vault:

Reviewer: ```suggestion
### Update the Consul on Kubernetes helm chart

Configure the Consul Helm chart to use the Server TLS certificates from Vault. 
```
Is that cleaner?

Author: Yes.  Thank you.  I'll replace it on several pages.
